src/producer/__tests__/idempotentProduceMessages.spec.js

```diff
@@ -0,0 +1,222 @@
+    )
+  })
+
+  it('sequential produce() calls > where produce() throws a retriable error, all messages are written to the partition once, in order', async () => {
+    for (const nodeId of [0, 1, 2]) {
+      const broker = await cluster.findBroker({ nodeId })
+
+      const brokerProduce = jest.spyOn(broker, 'produce')
+      brokerProduce.mockImplementationOnce(async () => {
+        await sleep(5)
+        throw new KafkaJSError('retriable error')
+      })
+    }
+
+    const messagesConsumed = []
+
+    for (const m of messages) {
+      await producer.send({ acks: -1, topic: topicName, messages: [m] })
+    }
+
+    await consumer.run({ eachMessage: async message => messagesConsumed.push(message) })
+
+    await waitForMessages(messagesConsumed, { number: messages.length })
+
+    messagesConsumed.forEach(({ message: { value } }, i) =>
+      expect(value.toString()).toEqual(`${i}`)
+    )
+  })
+
+  it('sequential produce() calls > where produce() throws a retriable error after the message is written to the log, all messages are written to the partition once, in order', async () => {
+    for (const nodeId of [0, 1, 2]) {
+      const broker = await cluster.findBroker({ nodeId })
+      const originalCall = broker.produce.bind(broker)
+      const brokerProduce = jest.spyOn(broker, 'produce')
+      brokerProduce.mockImplementationOnce()
+      brokerProduce.mockImplementationOnce()
+      brokerProduce.mockImplementationOnce(async (...args) => {
+        await originalCall(...args)
+        throw new KafkaJSError('retriable error')
+      })
+    }
+
+    const messagesConsumed = []
+
+    for (const m of messages) {
+      await producer.send({ acks: -1, topic: topicName, messages: [m] })
+    }
+
+    await consumer.run({ eachMessage: async message => messagesConsumed.push(message) })
+
+    await waitForMessages(messagesConsumed, { number: messages.length })
+
+    messagesConsumed.forEach(({ message: { value } }, i) =>
+      expect(value.toString()).toEqual(`${i}`)
+    )
+  })
+
+  it('concurrent produce() calls > all messages are written to the partition once', async () => {
+    const messagesConsumed = []
+
+    await Promise.all(
+      messages.map(m => producer.send({ acks: -1, topic: topicName, messages: [m] }))
+    )
+
+    await consumer.run({ eachMessage: async message => messagesConsumed.push(message) })
+
+    await waitForMessages(messagesConsumed, { number: messages.length })
+
+    expect(arrayUnique(messagesConsumed.map(({ message: { value } }) => value))).toHaveLength(
+      messages.length
+    )
+  })
+
+  it('concurrent produce() calls > where produce() throws a retriable error on the first call, all subsequent calls throw UNKNOWN_PRODUCER_ID', async () => {
+    for (const nodeId of [0, 1, 2]) {
+      const broker = await cluster.findBroker({ nodeId })
+
+      const brokerProduce = jest.spyOn(broker, 'produce')
+      brokerProduce.mockImplementationOnce(async () => {
+        await sleep(100)
```

Comment:
Do these sleeps make any meaningful difference? Our tests are already taking a bit, so if we don't need sleeps to test some timing thing, let's not add them.

Reply:
Yes, they were meaningful.  The intent was to make sure the other concurrent produce requests got initiated before the mocked one threw the error. But agree that's far from ideal way to do this.

I have now changed it to actually await on the number of times the produce request is called before throwing.